VMware: Refactor vmware_cluster into several modules

mariolenz opened the pull request:
##### SUMMARY
Refactor vmware_cluster into several modules (vmware_cluster, vmware_cluster_drs, vmware_cluster_ha and vmware_cluster_vsan) as discussed in #58023.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vmware_cluster

##### ADDITIONAL INFORMATION
vmware_cluster lacks a lot of configuration options for DRS, HA and vSAN. Implementing them all in vmware_cluster would make the module hard to maintain. Therefore, splitting it into several modules and implementing the missing configuration options in them seems a good idea to me.

This is step one, refactoring vmware_cluster into several modules. Step two, implementing more configuration options for DRS, HA and vSAN, will follow.

mariolenz committed:
Refactor vmware_cluster into several modules

mariolenz commented:
@Akasurde @goneri @jillr I rebased this PR to the current devel branch, still looks OK to me.

jillr commented on lib/ansible/modules/cloud/vmware/vmware_cluster.py:
```diff
@@ -360,11 +136,11 @@ def state_create_cluster(self):
         Destroy cluster
         """
-        changed, result = False, None
+        changed, result = True, None
```
```suggestion
        changed, result = False, None
```
`changed` should be False until an actual change action sets it to True.

jillr reviewed (commented):
Thanks for this patch @mariolenz.  On a first pass this looks generally reasonable, I'll try to give it a second pass next week.  
We'll need integration tests for the new modules, and a porting guide entry in `docs/docsite/rst/porting_guides/porting_guide_2.9.rst`.

mariolenz committed:
Update lib/ansible/modules/cloud/vmware/vmware_cluster.py

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

mariolenz commented:
> We'll need integration tests for the new modules

I haven't written any tests yet... I'll have to work on this. As far as I know, you use govcsim(?) for your tests and I'm not sure if this really simulates the whole vCenter API. I generally test against the vCSA in our test environment... but I'll try.

> porting guide entry in `docs/docsite/rst/porting_guides/porting_guide_2.9.rst`.

I'll have a look at this, too... give me a some time.

btw: In vmware_cluster_drs, vmware_cluster_ha and vmware_cluster_vsan there is the same code:

```
       try:
            self.datacenter = find_datacenter_by_name(self.content, self.datacenter_name)
            if self.datacenter is None:
                self.module.fail_json(msg="Datacenter %s does not exist." % self.datacenter_name)

            self.cluster = self.find_cluster_by_name(cluster_name=self.cluster_name)
            if self.cluster is None:
                self.module.fail_json(msg="Cluster %s does not exist." % self.cluster_name)

        except vmodl.RuntimeFault as runtime_fault:
            self.module.fail_json(msg=to_native(runtime_fault.msg))
        except vmodl.MethodFault as method_fault:
            self.module.fail_json(msg=to_native(method_fault.msg))
        except Exception as generic_exc:
            self.module.fail_json(msg="Failed to check configuration"
                                      " due to generic exception %s" % to_native(generic_exc))
```

I think this should be moved to lib/ansible/module_utils/vmware.py. Do you want me to do it in this PR or later on?

mariolenz committed:
Add integration tests for vmware_cluster_drs an vmware_cluster_ha

mariolenz commented:
@jillr I've written (very) basic integration tests for vmware_cluster_drs and vmware_cluster_ha. Unfortunately, govcsim seems very limited. There are some things I'd like to have tested (like idempotency) that don't work, although they work fine with a real vCenter.

I couldn't write a test for vmware_cluster_vsan, either, because govcsim doesn't really know about vSAN. I wasn't even able to check if it's enabled or not. I see a problem here: You want me to write an integration test, but I can't because govcsim doesn't simulate the required functionality. I could write one with `when: vcsim is not defined` if you like. But you'll need a _real_ vCenter to run them.

jillr commented:
@mariolenz  OK thanks for looking into that. We're actually in the process now of building a new CI environment that will test against a real VMware environment so we can skip the things vcsim doesn't cover now, or if you're willing to write them we can just leave those tests disabled until that comes fully online.

mariolenz committed:
Add porting guide entry

mariolenz committed:
Add integration test for vmware_cluster_vsan

mariolenz commented:
@jillr I've added an integration test for vmware_cluster_vsan that runs only `when: vcsim is not defined`. Just a thought: Maybe checking for missing integration tests is something that ansibot could do automatically, so you don't have to?

I've also added a porting guide entry. I didn't really understand the syntax but maybe I'm lucky and got it right just by chance.

jillr commented:
@goneri can you please take a look at the tests in this PR?

goneri commented:
@mariolenz, this is a major improvement. Thanks for the contribution. The patch changes `vmware_cluster` behaviour and thus breaks test/integration/targets/prepare_vmware_tests/tasks/setup_cluster.yml. Could you refresh this file too.

This doc explains how to run the test-suite locally: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst

mariolenz committed:
Update setup_cluster.yml

mariolenz commented:
@goneri Sorry, I havn't realized that I have broken something. I only tested against govcsim and setup_cluster.yml isn't executed in that case. I hope it's fixed now.

mariolenz commented:
@Akasurde @goneri @jillr Do you have any news for me on this PR? I hope I fixed all issues you've mentioned, but if you think otherwise or if there are some more don't hesitate to tell me. I'll do what I can.

jillr reviewed (commented):
Code LGTM, @acozine could you take a look at the porting guide here please?  We're breaking up a module into several smaller modules.

acozine commented on docs/docsite/rst/porting_guides/porting_guide_2.9.rst:
```diff
@@ -66,6 +66,7 @@ No notable changes
 Noteworthy module changes
 -------------------------
 
+* :ref:`vmware_cluster <vmware_cluster_module>` has been refactored into several specialized modules; the functionality to configure DRS, HA and vSAN has been moved to :ref:`vmware_cluster_drs <vmware_cluster_drs_module>`, :ref:`vmware_cluster_ha <vmware_cluster_ha_module>` and :ref:`vmware_cluster_vsan <vmware_cluster_vsan_module>`.
```
I'd add a short statement about the benefit to users (or to developers) of this change, and make it clear that the old module still exists. I don't know what the motivation for this change was, but I'll give an example - if the idea was to make maintenance easier, you could say:
```
* :ref:`vmware_cluster <vmware_cluster_module>` was refactored for easier maintenance/bugfixes. The original module only adds/removes clusters. Use the three new, specialized modules to configure clusters. Configure DRS with :ref:`vmware_cluster_drs <vmware_cluster_drs_module>`, HA with :ref:`vmware_cluster_ha <vmware_cluster_ha_module>`, and vSAN with :ref:`vmware_cluster_vsan <vmware_cluster_vsan_module>`.
```

mariolenz replied:
Yes, the idea was to make maintenance easier. I've copied your suggestion.

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster.py:
```diff
@@ -20,8 +20,7 @@
 description:
-    - This module can be used to add, remove and update VMware vSphere clusters and its configurations.
-    - Module can manage HA, DRS and VSAN related configurations.
+    - This module can be used to add and remove VMware vSphere clusters.
```
```suggestion
    - Adds or removes VMware vSphere clusters.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+DOCUMENTATION = r'''
+---
+module: vmware_cluster_drs
+short_description: Manage DRS on VMware vSphere clusters
```
Define DRS at least once:
```suggestion
short_description: Manage Distributed Resource Scheduler (DRS) on VMware vSphere clusters
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+module: vmware_cluster_drs
+short_description: Manage DRS on VMware vSphere clusters
+description:
+    - This module can be used to manage DRS on VMware vSphere clusters.
```
```suggestion
    - Manages DRS on VMware vSphere clusters.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+      aliases: [ datacenter_name ]
+    enable_drs:
+      description:
+      - If set to C(yes), will enable DRS when the cluster is created.
```
```suggestion
      - Whether to enable DRS when the cluster is created.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+      default: 'no'
+    drs_enable_vm_behavior_overrides:
+      description:
+      - Determines whether DRS Behavior overrides for individual virtual machines are enabled.
```
```suggestion
      - Whether DRS Behavior overrides for individual virtual machines are enabled.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+    drs_default_vm_behavior:
+      description:
+      - Specifies the cluster-wide default DRS behavior for virtual machines.
+      - If set to C(partiallyAutomated), then vCenter generate recommendations for virtual machine migration and
```
I'm not sure I have this right, but here's my guess at what this means:
```suggestion
      - If set to C(partiallyAutomated), vCenter generates recommendations for virtual machine migration and
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+      description:
+      - Specifies the cluster-wide default DRS behavior for virtual machines.
+      - If set to C(partiallyAutomated), then vCenter generate recommendations for virtual machine migration and
+        for the placement with a host. vCenter automatically implement placement at power on.
```
Best guess part two:
```suggestion
        for the placement with a host, then automatically implements placement and power recommendations.
```

mariolenz replied:
This is not really correct because vCenter will not make any power recommendations, only placement recommendations when the VM is powered on. But I've changed the description a bit, hope it's OK now.

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+      - Specifies the cluster-wide default DRS behavior for virtual machines.
+      - If set to C(partiallyAutomated), then vCenter generate recommendations for virtual machine migration and
+        for the placement with a host. vCenter automatically implement placement at power on.
+      - If set to C(manual), then vCenter generate recommendations for virtual machine migration and
```
```suggestion
      - If set to C(manual), then vCenter generates recommendations for virtual machine migration and
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+      - If set to C(partiallyAutomated), then vCenter generate recommendations for virtual machine migration and
+        for the placement with a host. vCenter automatically implement placement at power on.
+      - If set to C(manual), then vCenter generate recommendations for virtual machine migration and
+        for the placement with a host. vCenter should not implement the recommendations automatically.
```
```suggestion
        for the placement with a host, but does not implement the recommendations automatically.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+        for the placement with a host. vCenter automatically implement placement at power on.
+      - If set to C(manual), then vCenter generate recommendations for virtual machine migration and
+        for the placement with a host. vCenter should not implement the recommendations automatically.
+      - If set to C(fullyAutomated), then vCenter should automate both the migration of virtual machines
```
```suggestion
      - If set to C(fullyAutomated), then vCenter automates both the migration of virtual machines
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+DOCUMENTATION = r'''
+---
+module: vmware_cluster_ha
+short_description: Manage HA on VMware vSphere clusters
```
Again, define acronyms at least once:
```suggestion
short_description: Manage High Availability (HA) on VMware vSphere clusters
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+module: vmware_cluster_ha
+short_description: Manage HA on VMware vSphere clusters
+description:
+    - This module can be used to manage HA on VMware vSphere clusters.
```
```suggestion
    - Manages HA configuration on VMware vSphere clusters.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      aliases: [ datacenter_name ]
+    enable_ha:
+      description:
+      - If set to C(yes) will enable HA when the cluster is created.
```
```suggestion
      - Whether to enable HA when the cluster is created.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      default: 'no'
+    ha_host_monitoring:
+      description:
+      - Indicates whether HA restarts virtual machines after a host fails.
```
```suggestion
      - Whether HA restarts virtual machines after a host fails.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      version_added: 2.8
+    ha_vm_monitoring:
+      description:
+      - Indicates the state of virtual machine health monitoring service.
```
```suggestion
      - State of virtual machine health monitoring service.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      type: str
+      choices: ['vmAndAppMonitoring', 'vmMonitoringOnly', 'vmMonitoringDisabled']
+      default: 'vmMonitoringDisabled'
+      version_added: 2.8
```
Should we mark this as added in 2.8? This functionality in this module didn't exist in 2.8.

mariolenz replied:
Changed to 2.9.

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      description:
+      - The number of seconds for the virtual machine's heartbeats to stabilize after
+        the virtual machine has been powered on.
+      - This setting is only valid if C(ha_vm_monitoring) is set to, either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```
```suggestion
      - Valid only when I(ha_vm_monitoring) is set to either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      description:
+      - Maximum number of failures and automated resets allowed during the time
+       that C(ha_vm_max_failure_window) specifies.
+      - This setting is only valid if C(ha_vm_monitoring) is set to, either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```
```suggestion
      - Valid only when I(ha_vm_monitoring) is set to either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      description:
+      - The number of seconds for the window during which up to C(ha_vm_max_failures) resets
+        can occur before automated responses stop.
+      - This setting is only valid if C(ha_vm_monitoring) is set to, either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```
```suggestion
      - Valid only when I(ha_vm_monitoring) is set to either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      version_added: 2.8
+    ha_restart_priority:
+      description:
+      - Determines the preference that HA gives to a virtual machine if sufficient capacity is not available
```
```suggestion
      - Priority HA gives to a virtual machine if sufficient capacity is not available
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,317 @@
+      description:
+      - Determines the preference that HA gives to a virtual machine if sufficient capacity is not available
+        to power on all failed virtual machines.
+      - This setting is only valid if C(ha_vm_monitoring) is set to, either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```
```suggestion
      - Valid only if I(ha_vm_monitoring) is set to either C(vmAndAppMonitoring) or C(vmMonitoringOnly).
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_vsan.py:
```diff
@@ -0,0 +1,188 @@
+module: vmware_cluster_vsan
+short_description: Manage vSAN on VMware vSphere clusters
+description:
+    - This module can be used to manage vSAN on VMware vSphere clusters.
```
```suggestion
    - Manages vSAN on VMware vSphere clusters.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_vsan.py:
```diff
@@ -0,0 +1,188 @@
+DOCUMENTATION = r'''
+---
+module: vmware_cluster_vsan
+short_description: Manage vSAN on VMware vSphere clusters
```
```suggestion
short_description: Manages virtual storage area network (vSAN) configuration on VMware vSphere clusters
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_vsan.py:
```diff
@@ -0,0 +1,188 @@
+      aliases: [ datacenter_name ]
+    enable_vsan:
+      description:
+      - If set to C(yes) will enable vSAN when the cluster is created.
```
```suggestion
      - Whether to enable vSAN when the cluster is created.
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_vsan.py:
```diff
@@ -0,0 +1,188 @@
+      default: 'no'
+    vsan_auto_claim_storage:
+      description:
+      - Determines whether the VSAN service is configured to automatically claim local storage
```
```suggestion
      - Whether the VSAN service is configured to automatically claim local storage
```

acozine commented on lib/ansible/modules/cloud/vmware/vmware_cluster_vsan.py:
```diff
@@ -0,0 +1,188 @@
+        on VSAN-enabled hosts in the cluster.
+      type: bool
+      default: False
+      version_added: 2.8
```
Same question about `version_added` throughout the new modules.

mariolenz replied:
Changed to 2.9.

acozine reviewed (commented):
Thanks @mariolenz for making the docs thorough. I found a few ways to improve them even more.

mariolenz committed:
vmware_cluster: Update module description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_drs: Update short_description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_drs: Update description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_drs: Update documentation for parameter enable_drs

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_drs: Update documentation for drs_enable_vm_behavior_overrides

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_drs: Update documentation for drs_default_vm_behavior

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
Update lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
Update lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
Update lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update short_description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_vsan: Update description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_vsan: Update short_description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_vsan: Update description for parameter enable_vsan

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_vsan: Update description for parameter vsan_auto_claim_storage

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update description

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update description for parameter enable_ha

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update description for parameter ha_host_monitoring

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update description for parameter ha_vm_monitoring

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update description for ha_vm_min_uptime

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update documentation for ha_vm_max_failures

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update documentation for parameter ha_vm_max_failure_window

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update documentation for parameter ha_restart_priority

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_ha: Update documentation for parameter ha_restart_priority

Co-Authored-By: Alicia Cozine <879121+acozine@users.noreply.github.com>

mariolenz committed:
vmware_cluster_drs/ha/vsan: Remove "when creatd" from documentation

mariolenz committed:
vmware_cluster_drs/ha/vsan: Change version_added from 2.8 to 2.9

mariolenz committed:
Update porting guide

mariolenz committed:
vmware_cluster_drs: Update documentation for parameter drs_default_vm_behavior

mariolenz commented:
@acozine Thanks for helping me with the documentation. I've used most of your sugestions and hope it's OK now.
Btw: I've removed the "when the cluster is created" from the module descriptions because they're to be used after the cluster has been created. This has been copy&paset'd from the old module.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_cluster.py:
```diff
@@ -34,139 +33,14 @@
+      type: str
       required: yes
       aliases: [ datacenter_name ]
-    enable_drs:
```
I would say keep this code / functionalities as is and warn user when they try to use this. Providing correct module name for the given functionality.

This will not break existing playbooks and will allow users to migrate to new modules. Removing code abruptly will introduce chaos.

mariolenz replied:
OK, I understand what you mean. I've reverted the vmware_cluster module to the original version plus some deprecation warnings. But I don't like it, there are one or two things that somehow feel wrong. I do have an idea how to do this in a better way, but I'll have to think this out.

mariolenz replied:
I've added three new parameters: `ignore_drs`, `ignore_ha` and `ignore_vsan`. This way, people can get rid of the deprication warnings when using the new modules by making vmware_cluster ignore DRS, HA and VSAN configuration.

Since the default is `False`, these new parameters shouldn't break any existing playbooks.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_cluster.py:
```diff
@@ -236,104 +88,28 @@ def __init__(self, module):
-        Manage HA Configuration
-        Returns: Cluster DAS configuration spec
-
-        """
```
```python
            msg = 'Configuring cluster HA using vmware_cluster module is deprecated and will be removed in version 2.12. Please use vmware_cluster_ha module for the new functionality.'
            self.module.deprecate(msg, '2.12')

```

Akasurde replied:
Same for DRS and vSAN

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_cluster.py:
```diff
@@ -20,8 +20,7 @@
 description:
-    - This module can be used to add, remove and update VMware vSphere clusters and its configurations.
-    - Module can manage HA, DRS and VSAN related configurations.
+    - Adds or removes VMware vSphere clusters.
```
Add note stating that there are three new modules to handle this functionality.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+      - If set to C(True), overrides C(drs_default_vm_behavior).
+      type: bool
+      default: True
+      version_added: 2.9
```
Not required for new module.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,217 @@
+        self.cluster = None
+
+        try:
+            self.datacenter = find_datacenter_by_name(self.content, self.datacenter_name)
```
These APIs does not raise anything so, try...except can be removed.

Akasurde reviewed (changes requested):
Rest LGTM

Akasurde commented:
@mariolenz Thanks for the contribution. Sorry, was not able to review this code before, but better late than never.

mariolenz committed:
Remove "version_added: 2.9" from parameters

Not required for new module.

mariolenz committed:
Remove "version_added: 2.9" from parameters

Not required for new module.

mariolenz committed:
Remove "version_added: 2.9" from parameters

Not required for new module.

goneri commented on test/integration/targets/vmware_cluster_vsan/tasks/main.yml:
```diff
@@ -0,0 +1,64 @@
+    - name: Ensure vSAN is disabled
+      assert:
+        that:
+            - "{{ cluster__result_0002.changed == true }}"
```
It should be `cluster_vsan_result_0002`

goneri reviewed (changes requested):
Looks great to me. just a minor comment. Thank you @mariolenz for the contribution.

mariolenz committed:
Fix C&P error in variable name

goneri commented on lib/ansible/modules/cloud/vmware/vmware_cluster_vsan.py:
```diff
@@ -0,0 +1,187 @@
+- name: Enable vSAN and claim storage automatically
+  vmware_cluster_vsan:
+    hostname: "{{ vcenter_server }}"
+    username: "{{ vcenter_user }}"
```
it should be:

```yaml
    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
```

goneri commented on lib/ansible/modules/cloud/vmware/vmware_cluster_drs.py:
```diff
@@ -0,0 +1,214 @@
+  vmware_cluster_drs:
+    hostname: "{{ vcenter_server }}"
+    username: "{{ vcenter_user }}"
+    password: "{{ vcenter_pass }}"
```
It should be:
```yaml
    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
```

goneri commented on lib/ansible/modules/cloud/vmware/vmware_cluster_ha.py:
```diff
@@ -0,0 +1,308 @@
+  vmware_cluster_ha:
+    hostname: "{{ vcenter_server }}"
+    username: "{{ vcenter_user }}"
+    password: "{{ vcenter_pass }}"
```
It should be:
```yaml
    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
```

mariolenz committed:
Remove unnecessary try...except in new vmware_cluster_* modules

mariolenz committed:
Restore old vmware_cluster functionality

mariolenz committed:
Improve documentation of new vmware_cluster_* modules

mariolenz committed:
New parameters to ignore DRS, HA and VSAN config

mariolenz committed:
Fix C&P error

mariolenz committed:
Document deprecation of DRS, HA and VSAN in vmware_cluster

mariolenz committed:
Document deprecation of DRS, HA and VSAN in vmware_cluster

goneri commented:
Hi @mariolenz , I tried to rebase your branch against `devel`. There is a little conflict in `docs/docsite/rst/porting_guides/porting_guide_2.9.rst`.

goneri commented on test/integration/targets/vmware_cluster_vsan/tasks/main.yml:
```diff
@@ -0,0 +1,64 @@
+    - name: Ensure vSAN is disabled
+      assert:
+        that:
+            - "{{ cluster_result_0002.changed == true }}"
```
It should be `cluster_vsan_result_0002`.changed, not `cluster_result_0002.changed`.

mariolenz replied:
Sorry, I didn't spot that. There were two underscores in the variable name, I thought you meant that.

mariolenz committed:
Fix yet another C&P error

mariolenz committed:
deprications -> deprecated_features

mariolenz committed:
Fix variable name

mariolenz commented:
@Akasurde @goneri Sorry for the many commits today, it's just too hot here to concentrate. But I think I'm through now and hope everything is OK.

@acozine Can you please have another look at the documentation? I've put the original functionality back into vmware_cluster, but marked it as deprecated in favor of the new modules.

mariolenz commented:
@acozine @Akasurde @goneri Sorry for the last force-push, but there has been a merge conflict with PR #58589. There haven't been any changes except resolving this merge conflict in `docs/docsite/rst/porting_guides/porting_guide_2.9.rst`. (If I didn't make any stupid mistake, that is...)

goneri commented:
> @acozine @Akasurde @goneri Sorry for the last force-push, but there has been a merge conflict with PR #58589. There haven't been any changes except resolving this merge conflict in `docs/docsite/rst/porting_guides/porting_guide_2.9.rst`. (If I didn't make any stupid mistake, that is...)

Thanks for the rebase. I personnally force-push all the time to keep my patchset clean and squash all the 'fix previous fix' commits. I think it's fine as soon as it's a private branch.

goneri reviewed (approved):
Thank you @mariolenz for this large refactoring :heart:.
